prow/cmd/webhook-server/main.go

```diff
@@ -48,47 +47,125 @@ type ClientInterface interface {
 	flag.Var(&dns, "dns", "DNS Names CA-Cert config")
 	flag.Parse()
-	exp := *e
-	projectId := *p
-	if len(projectId) == 0 {
-		logrus.Fatal("project-id flag not supplied")
+	o := gatherOptions(flag.NewFlagSet(os.Args[0], flag.ExitOnError), os.Args[1:]...)
+	o.dnsNames = dns.Strings()
+	if err := o.Validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid options")
 	}
-	client, err := secretmanager.NewClient(projectId, false)
+	kubeCfg, err := o.kubernetes.InfrastructureClusterConfig(o.dryRun)
 	if err != nil {
-		logrus.WithError(err).Fatal("Unable to create secret manager client")
+		logrus.WithError(err).Fatal("Error getting kubeconfig")
 	}
+	var certFile string
+	var privKeyFile string
+	var tempDir string
 	ctx := context.Background()
-	cert, privKey, err := getGCPSecrets(client, ctx, exp, dns)
+	cl, err := ctrlruntimeclient.New(kubeCfg, ctrlruntimeclient.Options{})
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not create writer client")
+	}
+	var client ClientInterface
+	if len(o.projectId) > 0 {
+		client, err = secretmanager.NewClient(o.projectId, false)
+		if err != nil {
+			logrus.WithError(err).Fatal("Unable to create secret manager client")
+		}
+		tempDir, certFile, privKeyFile = handleSecrets(client, ctx, o, cl)
+		defer os.RemoveAll(tempDir)
+	}
+	if len(o.fileSystemPath) > 0 {
+		absPath, err := filepath.Abs(o.fileSystemPath)
+		if err != nil {
+			logrus.WithError(err).Fatal("Unable to generate absolute file path")
+		}
+		client = &FileSystemClient{
+			path:   absPath,
+			expiry: o.expiryInYears,
+			dns:    o.dnsNames,
+		}
+		tempDir, certFile, privKeyFile = handleSecrets(client, ctx, o, cl)
+		defer os.RemoveAll(tempDir)
+	}
+	http.HandleFunc("/validate", serveValidate)
+	logrus.Info("Listening on port 8008...")
+	logrus.Fatal(http.ListenAndServeTLS(":8008", certFile, privKeyFile, nil))
+}
+
+func handleSecrets(client ClientInterface, ctx context.Context, o options, cl ctrlruntimeclient.Client) (string, string, string) {
+	cert, privKey, _, err := getGCPSecrets(client, ctx, o.expiryInYears, o.dnsNames)
+	var caPem string
 	if err != nil && strings.Contains(err.Error(), gcpSecretError) {
-		logrus.Infof("%v, Will now proceed to create GCP Secret", err)
-		cert, privKey, err = createGCPSecret(client, ctx, exp, dns.Strings())
+		logrus.Infof("%v, Will now proceed to create secrets", err)
+		cert, privKey, caPem, err = createGCPSecret(client, ctx, o.expiryInYears, o.dnsNames)
 		if err != nil {
 			logrus.WithError(err).Fatal("Unable to create ca certificate")
 		}
+		err := createValidatingWebhookConfig(ctx, caPem, cl)
+		if err != nil {
+			logrus.WithError(err).Fatalf("Unable to generate ValidationWebhookConfig")
+		}
 	} else if err != nil {
-		logrus.WithError(err).Fatal("Unable to get GCP Secret")
+		logrus.WithError(err).Fatal("Unable to get secrets")
 	}
 	if err = isCertValid(cert); err != nil {
```

Comment:
`isCertValid` should return `(bool, error)` imo:
- if anything unexpected happened, for example network flakiness , return `nil, error`
- if the cert is valid, return `true, nil`
- if the cert is invalid, return `false, nil`

Reply:
was doing this previously in the first ever PR as you suggested but cole suggested otherwise

Reply:
was not aware of that, was it because `isCertValid` would never return unexpected error? if that's the case then it should only return `bool` instead of `error`

Reply:
yeah it also will never return network fail. It simply checks the time field in the certificate struct

Reply:
I see, imo it makes more sense to make it return bool, as cert expired qualifies as `isCertValid -> false`